sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -801,22 +807,84 @@ checkType type_ err = do
     sameType <- isUpgradedType type_
     unless sameType $ diagnosticWithContextF present' err
 
+checkQualVal :: UpgradingDeps -> Upgrading (Qualified ExprValName) -> Bool
+checkQualVal deps name =
+    let namesAreSame = foldU (==) (fmap removePkgId name)
+        qualificationIsSame = foldU (==) (fmap qualPackage name)
+        qualificationIsUpgraded =
+          case qualPackage <$> name of
+            Upgrading { _past = PRImport pastPkgId, _present = PRImport presentPkgId } ->
+              pkgIdsAreUpgraded deps Upgrading { _past = pastPkgId, _present = presentPkgId }
+            _ -> False
+    in
+    namesAreSame && (qualificationIsSame || qualificationIsUpgraded)
+  where
+    pkgIdsAreUpgraded :: UpgradingDeps -> Upgrading PackageId -> Bool
+    pkgIdsAreUpgraded deps pkgId =
+      case HMS.lookup <$> pkgId <*> pure deps of
+        Upgrading { _past = Just (pastName, mbPastVersion, pastSupportsUpgrades, pastIsUtilityPkg), _present = Just (presentName, mbPresentVersion, presentSupportsUpgrades, presentIsUtilityPkg) } ->
+          pastName == presentName &&
+          case (pastSupportsUpgrades, presentSupportsUpgrades) of
+            (False, False) -> foldU (==) pkgId -- Both packages don't support upgrades (<= LF 1.15) so IDs should be identical
+            (True, True) ->
+              case (pastIsUtilityPkg, presentIsUtilityPkg) of
+                (True, True) ->
+                  case (mbPastVersion, mbPresentVersion) of
+                    (Just pastVersion, Just presentVersion) -> pastVersion < presentVersion
+                    _ -> True -- This case should not be possible
+                (False, False) ->
+                  presentName `elem` [PackageName "daml-stdlib", PackageName "daml-prim"]
+                  || case (mbPastVersion, mbPresentVersion) of
+                      (Just pastVersion, Just presentVersion) -> pastVersion < presentVersion
+                      (Nothing, Nothing) -> True -- Only daml-prim has no version field
+                      _ -> True -- This case should not be possible
+                _ -> False -- Shouldn't go from utility package to non-utility package
+            _ -> False -- 
+        _ ->
+          False
```

Comment:
This shouldn't be possible, fine to leave til next PR

Reply:
https://github.com/digital-asset/daml/pull/19970